### PR TITLE
feat: personal memos page

### PR DESCRIPTION
**Pull request by hyoban**
showing memos without logging in can let us use it like a personal Twitter.

### CODE REVIEW

**Comment by boojack** [comment]
I think it's better to create **a new page** to show someone's **public** memos. Just like https://jike.city/xidong, wdyt?

**Comment by hyoban** [comment]
My idea is to make it like Github, others can see what's public by default, and at the same time, we can hide some memos.

**Comment by boojack** [comment]
But the current main page UI isn't good for visitor because of the editor and some action buttons. A new pure page to display public memos and some extra information might be better for viewer.

**Comment by boojack** [comment]
> But the current main page UI isn't good for visitor because of the editor and some action buttons. A new pure page to display public memos and some extra information might be better for viewer.

After taking a deep look, I think it's good for viewing when hiding some extra elements.

image

**Comment by boojack** [comment]
LGTM. Another question: how do we view other member's memos?

**Comment by hyoban** [comment]
This is also a more troublesome issue that I have considered. At present, I let it display the owner's memos by default.

**Comment by hyoban** [comment]
I think most people will choose to build their own memos. 
Allowing multiple users to be created doesn't seem very useful.

**Comment by hyoban** [comment]
ok, I'll try to handle each user's homepage

**Comment by hyoban** [comment]
@boojack maybe you can review the code when you are free

File: server/user.go [hunk 1]
```diff
@@ -51,6 +51,26 @@ func (s *Server) registerUserRoutes(g *echo.Group) {
 		return nil
 	})
 
+	g.GET("/username/:id", func(c echo.Context) error {
```
Comment (boojack) [inline server/user.go:54]: `/user/:id` might be better

File: server/user.go [hunk 2]
```diff
@@ -51,6 +51,26 @@ func (s *Server) registerUserRoutes(g *echo.Group) {
...
+
+		user, err := s.Store.FindUser(&api.UserFind{
+			ID: &id,
+		})
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to fetch user").SetInternal(err)
+		}
+
+		c.Response().Header().Set(echo.HeaderContentType, echo.MIMEApplicationJSONCharsetUTF8)
+		if err := json.NewEncoder(c.Response().Writer).Encode(composeResponse(user.Name)); err != nil {
```
Comment (boojack) [inline server/user.go:68]: Need to check the `user` is `nil` firstly.

File: store/db/seed/10001__user.sql [hunk 3]
```diff
@@ -17,3 +17,23 @@ VALUES
...
+VALUES
+  (
+    102, 
+    'demo2@usememos.com',
+    'USER',
+    'Demo User',
+    'demo_open_id2',
+    -- raw password: secret
+    '$2a$14$ajq8Q7fbtFRQvXpdCq7Jcuy.Rx1h/L4J60Otx.gyNLbAYctGMJ9tK'
+  );
```
Comment (boojack) [inline store/db/seed/10001__user.sql:39]:
```suggestion
  (
    102, 
    'jack@usememos.com',
    'USER',
    'Jack',
    'jack_open_id',
    -- raw password: secret
    '$2a$14$ajq8Q7fbtFRQvXpdCq7Jcuy.Rx1h/L4J60Otx.gyNLbAYctGMJ9tK'
  );
```

File: web/src/components/Memo.tsx [hunk 4]
```diff
@@ -10,6 +10,7 @@ import showMemoCardDialog from "./MemoCardDialog";
 import showShareMemoImageDialog from "./ShareMemoImageDialog";
 import toastHelper from "./Toast";
 import "../less/memo.less";
+import { useAppSelector } from "../store";
```
Comment (boojack) [inline web/src/components/Memo.tsx:13]: The import order needs update, `.less` should be put in last.

File: web/src/helpers/api.ts [hunk 5]
```diff
@@ -44,16 +44,20 @@ export function getUserList() {
   return axios.get<ResponseObject<User[]>>("/api/user");
 }
 
+export function getUserNameById(id: number) {
```
Comment (boojack) [inline web/src/helpers/api.ts:47]: We can just write a `getUserById`, maybe some fields of user might be useful later not only his name.

File: web/src/helpers/api.ts [hunk 6]
```diff
@@ -44,16 +44,20 @@ export function getUserList() {
...
+  return axios.get<ResponseObject<string>>(`/api/username/${id}`);
+}
+
 export function patchUser(userPatch: UserPatch) {
   return axios.patch<ResponseObject<User>>("/api/user/me", userPatch);
 }
 
-export function getMemoList() {
-  return axios.get<ResponseObject<Memo[]>>("/api/memo");
+export function getMemoList(userID?: number) {
```
Comment (boojack) [inline web/src/helpers/api.ts:55]: In frontend, the id should be named in `camelCase`. So it should be `userId: UserId`.

File: web/src/helpers/api.ts [hunk 7]
```diff
@@ -44,16 +44,20 @@ export function getUserList() {
...
+  return axios.get<ResponseObject<Memo[]>>(`/api/memo?userID=${userID ? userID : ""}`);
```
Comment (boojack) [inline web/src/helpers/api.ts:56]:
```suggestion
  return axios.get<ResponseObject<Memo[]>>(`/api/memo${userID ? "?userID=" + userID : ""}`);
```

File: web/src/components/Memo.tsx [hunk 8]
```diff
@@ -158,38 +160,40 @@ const Memo: React.FC<Props> = (props: Props) => {
...
-                  <span className="tip-text">{memo.pinned ? "Unpin" : "Pin"}</span>
-                </div>
-                <div className="btn" onClick={handleEditMemoClick}>
-                  <img className="icon-img" src="/icons/edit.svg" alt="" />
-                  <span className="tip-text">Edit</span>
-                </div>
-                <div className="btn" onClick={handleGenMemoImageBtnClick}>
-                  <img className="icon-img" src="/icons/share.svg" alt="" />
-                  <span className="tip-text">Share</span>
+        {user && (
```
Comment (boojack) [inline web/src/components/Memo.tsx:163]: We have a `userService` with some common functions. So I think we can use it to check the userId(in param) and the current userId(in backend session) is same or different. For example, create a new function `isVisitor(): boolean` in `userService` that will check the login and visitor status. With it, these checks will looks better.

File: web/src/components/MemoList.tsx [hunk 9]
```diff
@@ -79,11 +79,26 @@ const MemoList: React.FC<Props> = () => {
...
+    const userId = !isNaN(Number(path)) ? Number(path) : undefined;
     memoService
-      .fetchAllMemos()
-      .then(() => {
-        setFetchStatus(false);
-        memoService.updateTagsState();
+      .fetchAllMemos(userId)
+      .then((data) => {
+        if (data.length === 0) {
+          locationService.replaceHistory("/");
```
Comment (boojack) [inline web/src/components/MemoList.tsx:88]: Why needs to redirect to `/`?

File: web/src/components/MemosHeader.tsx [hunk 10]
```diff
@@ -30,7 +30,9 @@ const MemosHeader: React.FC<Props> = () => {
     const now = Date.now();
     if (now - prevRequestTimestamp > 10 * 1000) {
       prevRequestTimestamp = now;
-      memoService.fetchAllMemos().catch(() => {
+      const path = locationService.getState().pathname.slice(1);
+      const userId = !isNaN(Number(path)) ? Number(path) : undefined;
```
Comment (boojack) [inline web/src/components/MemosHeader.tsx:34]: Do these check in `userService`.

File: server/user.go (hunk 1 above)
Comment (boojack) [inline server/user.go:54]: GET user by id would be more useful.

**Review by boojack — CHANGES_REQUESTED** [review]

File: server/user.go (hunk 1 above)
Comment (boojack) [reply server/user.go:54]: GET user by id would be more useful. But needs to remove the some sensitize fields. e.g. `OpenID`

Comment (hyoban) [reply server/user.go:54]: That's what I named it at first, but most of the user's information shouldn't be made public

Comment (boojack) [reply server/user.go:54]: Or `/user/:id/name`?

File: server/memo.go [hunk 11]
```diff
@@ -60,7 +60,26 @@ func (s *Server) registerMemoRoutes(g *echo.Group) {
 	})
 
 	g.GET("/memo", func(c echo.Context) error {
-		userID := c.Get(getUserIDContextKey()).(int)
+		userID, ok := c.Get(getUserIDContextKey()).(int)
+		if !ok {
+			if c.QueryParam("userID") != "" {
+				var err error
+				userID, err = strconv.Atoi(c.QueryParam("userID"))
```
Comment (boojack) [inline server/memo.go:67]:
```suggestion
				userID, err := strconv.Atoi(c.QueryParam("userID"))
```

File: server/memo.go [hunk 12]
```diff
@@ -60,7 +60,26 @@ func (s *Server) registerMemoRoutes(g *echo.Group) {
...
+				}
+			} else {
+				ownerUserType := api.Owner
+				ownerUser, err := s.Store.FindUser(&api.UserFind{
+					Role: &ownerUserType,
+				})
+				if err != nil {
+					return echo.NewHTTPError(http.StatusInternalServerError, "Failed to find owner user").SetInternal(err)
+				}
+				userID = ownerUser.ID
```
Comment (boojack) [inline server/memo.go:79]: needs to check ownerUser is whether `nil`

File: server/user.go [hunk 13]
```diff
@@ -51,6 +51,26 @@ func (s *Server) registerUserRoutes(g *echo.Group) {
...
+	g.GET("/user/:id/name", func(c echo.Context) error {
+		id, err := strconv.Atoi(c.Param("id"))
+		if err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, "Malformatted user id").SetInternal(err)
+		}
+
+		user, err := s.Store.FindUser(&api.UserFind{
+			ID: &id,
+		})
+		if err != nil || user == nil {
```
Comment (boojack) [inline server/user.go:63]: if user==nil, we should not return `http.StatusInternalServerError` but `user not found`.

File: web/src/components/Sidebar.tsx [hunk 14]
```diff
@@ -8,6 +8,7 @@ import UsageHeatMap from "./UsageHeatMap";
 import ShortcutList from "./ShortcutList";
 import TagList from "./TagList";
 import "../less/siderbar.less";
+import { userService } from "../services";
```
Comment (boojack) [inline web/src/components/Sidebar.tsx:11]: The import order needs update.

File: web/src/helpers/api.ts [hunk 15]
```diff
@@ -44,16 +44,20 @@ export function getUserList() {
...
-export function getMemoList() {
-  return axios.get<ResponseObject<Memo[]>>("/api/memo");
+export function getMemoList(userId?: number) {
+  return axios.get<ResponseObject<Memo[]>>(`/api/memo${userId ? "?userID=" + userId : ""}`);
 }
 
-export function getArchivedMemoList() {
-  return axios.get<ResponseObject<Memo[]>>("/api/memo?rowStatus=ARCHIVED");
+export function getArchivedMemoList(userId?: number) {
+  return axios.get<ResponseObject<Memo[]>>(`/api/memo?rowStatus=ARCHIVED${userId ? "?userID=" + userId : ""}`);
```
Comment (boojack) [inline web/src/helpers/api.ts:60]:
```suggestion
  return axios.get<ResponseObject<Memo[]>>(`/api/memo?rowStatus=ARCHIVED${userId ? "&userID=" + userId : ""}`);
```

File: web/src/services/memoService.ts [hunk 16]
```diff
@@ -10,21 +11,26 @@ const convertResponseModelMemo = (memo: Memo): Memo => {
   };
 };
 
+const getUserIdFromPath = () => {
```
Comment (boojack) [inline web/src/services/memoService.ts:14]: Put this function into userService might be better

**Review by boojack — CHANGES_REQUESTED** [review]

File: server/memo.go (hunk 11 above)
Comment (hyoban) [reply server/memo.go:67]: I can't modify it like this, the userID will be overwritten in the if area

File: web/src/services/memoService.ts [hunk 17]
```diff
@@ -10,21 +11,26 @@ const convertResponseModelMemo = (memo: Memo): Memo => {
...
+  const path = locationService.getState().pathname.slice(1);
```
Comment (boojack) [inline web/src/services/memoService.ts:15]: I think we could update the route to `/u/:userId`. It's more clearly, and we could create a new route `/m/:memoId` for viewing a memo with id in the future.

File: server/memo.go (hunk 11 above)
Comment (boojack) [reply server/memo.go:67]: Sorry, my bad

File: web/src/services/memoService.ts (hunk 17 above)
Comment (boojack) [reply web/src/services/memoService.ts:15]: And use a regex `/^\/u\/(\d+)/` to match the `:userId`

Comment (boojack) [reply web/src/services/memoService.ts]: My meaning is about the frontend route not the backend API. The current looks like is `/:userId`, and the `/u/:userId` will be more suitable.

Comment (hyoban) [reply web/src/services/memoService.ts]: sorry, i will correct it soon

**Comment by hyoban** [comment]
@boojack maybe we need an eslint config for imports sort, or [eslint-plugin-import](https://github.com/import-js/eslint-plugin-import)

File: web/src/components/UserBanner.tsx [hunk 18]
```diff
@@ -1,26 +1,48 @@
-import { useCallback, useState } from "react";
+import { useCallback, useEffect, useState } from "react";
 import { useAppSelector } from "../store";
 import { locationService } from "../services";
 import MenuBtnsPopup from "./MenuBtnsPopup";
 import "../less/user-banner.less";
+import * as api from "../helpers/api";
+import toastHelper from "./Toast";
```
[the remark this answers is not in the document]
Comment (hyoban) [reply web/src/components/UserBanner.tsx]: maybe we need an eslint config for imports sort, or [eslint-plugin-import](https://github.com/import-js/eslint-plugin-import)

[the remark this answers is not in the document]
Comment (boojack) [reply web/src/components/UserBanner.tsx]: LGTM, maybe you can use `yarn lint` in your local env.

[the remark this answers is not in the document]
Comment (hyoban) [reply web/src/components/UserBanner.tsx]: Personally, I will use [@antfu/eslint-config](https://github.com/antfu/eslint-config)

[the remark this answers is not in the document]
Comment (hyoban) [reply web/src/components/UserBanner.tsx]: if we only use eslint config, we can not auto-fix the order

[the remark this answers is not in the document]
Comment (boojack) [reply web/src/components/UserBanner.tsx]: I will take a look at eslint.

**Comment by hyoban** [comment]
This is my first contribution to a little feat, I have learned a lot, thank you

**Pull request merged**